packages/swarm-components/src/Dropdown.jsx

```diff
@@ -0,0 +1,462 @@
+					if (event.key === 'ArrowDown') {
+						event.preventDefault(); // prevent scroll
+						setState({ ...state, ...openAtFirstItem() });
+					} else if (event.key === 'ArrowUp') {
+						event.preventDefault(); // prevent scroll
+						setState({ ...state, ...openAtFirstItem() });
+					}
+				})}
+				{...props}
+			/>
+		);
+	}
+);
+
+MenuButton.propTypes = {
+	onClick: func,
+	onKeyDown: func,
+	children: node,
+};
+
+let MenuItem = React.forwardRef(
+	(
+		{
+			onSelect,
+			onClick,
+			role = 'menuitem',
+			index,
+			onKeyDown,
+			onMouseMove,
+			onMouseLeave,
+			setState: propSetState,
+			state: propState,
+			_ref,
+			...rest
+		},
+		ref
+	) => {
+		const { state, setState } = useContext(MenuContext);
+		let isSelected = state && index === state.selectionIndex;
+		let select = () => {
+			onSelect();
+			setState({ ...state, ...close() });
+		};
+
+		const itemRef = useRef(null);
+
+		// assignRef(ref, node);
+		// assignRef(_ref, node);
+
+		useEffect(() => {
+			if (itemRef.current && state.selectionIndex === index) {
+				itemRef.current.focus();
+			}
+		}, [state.selectionIndex]);
+		return (
+			<div
+				{...rest}
+				ref={itemRef}
+				data-reach-menu-item={role === 'menuitem' ? true : undefined}
+				role={role}
+				tabIndex="-1"
+				data-selected={role === 'menuitem' && isSelected ? true : undefined}
+				onClick={wrapEvent(onClick, event => {
+					select();
+				})}
+				onKeyDown={wrapEvent(onKeyDown, event => {
+					if (event.key === 'Enter') {
+						// prevent the button from being "clicked" by
+						// this "Enter" keydown
+						event.preventDefault();
+						select();
+					}
+				})}
+				onMouseMove={wrapEvent(onMouseMove, event => {
+					if (!isSelected) {
+						setState({ ...state, ...selectItemAtIndex(index) });
+					}
+				})}
+				onMouseLeave={wrapEvent(onMouseLeave, event => {
+					// clear out selection when mouse over a non-menu item child
+					setState({ ...state, selectionIndex: -1 });
+				})}
+			/>
+		);
+	}
+);
+
+MenuItem.propTypes = {
+	onSelect: func.isRequired,
+	onClick: func,
+	role: string,
+	state: object,
+	setState: func,
+	index: number,
+	onKeyDown: func,
+	onMouseMove: func,
+	_ref: func,
+};
+
+let k = () => {};
+
+////////////////////////////////////////////////////////////////////////
+let MenuLink = React.forwardRef(
+	(
+		{
+			onKeyDown,
+			onClick,
+			component: Comp,
+			as: AsComp = 'a',
+			style,
+			setState,
+			state,
+			index,
+			_ref,
+			...props
+		},
+		ref
+	) => {
+		const Link = Comp || AsComp;
+		if (Comp) {
+			console.warn(
+				'[@reach/menu-button]: Please use the `as` prop instead of `component`.'
+			);
+		}
+		return (
+			<MenuItem role="none" index={index} onSelect={k} _ref={k}>
+				<Link
+					role="menuitem"
+					data-reach-menu-item
+					tabIndex="-1"
+					data-selected={index === state.selectionIndex ? true : undefined}
+					onClick={wrapEvent(onClick, event => {
+						setState({ ...state, ...close() });
```

Comment:
no need to spread `...state` in this `setState` call, just `setState(close())`

Reply:
spreading `state` is necessary in hooks updater functions or it will result in erasing the other attributes being set.